compiler/rustc_ast/src/tokenstream.rs

```diff
@@ -265,10 +265,12 @@ impl AttrTokenStream {
                             "Failed to find trailing delimited group in: {target_tokens:?}"
                         );
                     }
-                    let mut flat: SmallVec<[_; 1]> = SmallVec::new();
+                    let size_hint =
+                        outer_attrs.iter().map(|attr| attr.tokens().len()).sum::<usize>()
```

Comment:
`attr.tokens()` is not a cheap operation, so calculating size hint like this is likely not beneficial.

Reply:
To optimize this I suggest actually looking at the number distribution for `flat.len()` and `target_tokens.len()` (e.g. by opening a raw file descriptor in append mode right here and dumping all the numbers into it), and set the `SmallVec`'s capacity and inline capacity based on that.

Even after that the optimization may be too small to be observable.

Reply:
Ok, then I suggest replacing `SmallVec` with `Vec` and reserving the initial capacity to `outer_attrs.len() + target_tokens.len()`.

Reply:
Sorry, I didn't notice `SmallVec` was used for compatibility with two other much more often executed branches `AttrTokenTree::{Token,Delimited}`.
Considering that it did make sense.